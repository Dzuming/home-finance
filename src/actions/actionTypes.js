const asyncActionType = type => ({
  PENDING: `${type}_PENDING`,
  SUCCESS: `${type}_SUCCESS`,
  ERROR: `${type}_ERROR`
});

const crudActionType = type => ({
  CREATE: `CREATE_${type}`,
  READ: `READ_${type}`,
  UPDATE: `UPDATE_${type}`,
  DELETE: `DELETE_${type}`
});

const draggedElements = type => ({
  READ: `READ_${type}`,
  SELECT: `SELECT${type}`,
  RESET: `RESET_${type}`,
  REMOVE: `REMOVE_${type}`
});

export const OPEN_SIDENAV = 'OPEN_SIDENAV';
export const AUTH_USER = 'AUTH_USER';
export const AUTH_REQUEST = 'AUTH_REQUEST';
export const REQUEST_LOGOUT = 'REQUEST_LOGOUT';
export const SUCCESS_LOGOUT = 'SUCCESS_LOGOUT';
export const SET_USER = 'SET_USER';
export const SET_DATE = 'SET_DATE';
export const API_REQUEST_GET = 'API_REQUEST_GET';
export const SET_CATEGORIES = 'SET_CATEGORIES';
export const API_REQUEST_POST = 'API_REQUEST_POST';
export const API_REQUEST_DELETE = 'API_REQUEST_DELETE';
export const API_REQUEST_PUT = 'API_REQUEST_PUT';
export const API_REQUEST = asyncActionType('API_REQUEST');
export const SPENDING_CRUD = crudActionType('SPENDING');
export const PROFIT_CRUD = crudActionType('PROFIT');
export const BUDGET_CRUD = crudActionType('BUDGET');
export const REVENUE_CRUD = crudActionType('REVENUE');
export const ASSUMPTION_CRUD = crudActionType('ASSUMPTION');
export const OVERALL_ASSUMPTIONS_CRUD = crudActionType('OVERALL_ASSUMPTIONS');
export const ASSUMPTION_TYPES = draggedElements('ASSUMPTION_TYPES');
export const CATEGORIES_DRAG = draggedElements('CATEGORIES');
