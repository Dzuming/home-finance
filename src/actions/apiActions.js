import * as types from './actionTypes';

export const apiRequestPending = () => ({
  type: types.API_REQUEST.PENDING,
});
export const apiRequestSuccess = () => ({
  type: types.API_REQUEST.SUCCESS,
});
