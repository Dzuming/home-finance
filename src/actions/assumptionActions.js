import * as types from './actionTypes';
import { createAction } from 'redux-actions';
import env from '../../environments/config';

export const fetchAssumptions = period => (dispatch, getState) => {
  const { id } = getState().user;
  return dispatch({
    type: types.API_REQUEST_GET,
    payload: {
      url: `${env.api_url}/api/assumptions/${id}/${period}`,
      success: setAssumption,
    },
  });
};

export const fetchOverallAssumptions = () => (dispatch, getState) => {
  const { id } = getState().user;
  return dispatch({
    type: types.API_REQUEST_GET,
    payload: {
      url: `${env.api_url}/api/assumptions/${id}`,
      success: setOverallAssumptions,
    },
  });
};

export const fetchAssumptionTypes = period => (dispatch, getState) => {
  const { id } = getState().user;
  return dispatch({
    type: types.API_REQUEST_GET,
    payload: {
      url: `${env.api_url}/api/assumptionTypes/${id}/${period}`,
      success: setAssumptionTypes,
    },
  });
};

export const setAssumption = createAction(types.ASSUMPTION_CRUD.READ);

export const setOverallAssumptions = createAction(
  types.OVERALL_ASSUMPTIONS_CRUD.READ,
);

export const setAssumptionTypes = createAction(
  types.ASSUMPTION_TYPES_DRAG.READ,
);

export const createAssumption = createAction(types.API_REQUEST_POST, data => ({
  url: `${env.api_url}/api/assumptions`,
  success: addAssumption,
  data,
}));

export const addAssumption = createAction(
  types.ASSUMPTION_CRUD.CREATE,
  assumption => ({
    assumption,
  }),
);

export const reduceAssumptionTypes = createAction(
  types.ASSUMPTION_TYPES_DRAG.REDUCE,
  type => ({
    type,
  }),
);

export const reduceCategories = createAction(
  types.CATEGORIES_DRAG.REDUCE,
  type => ({
    type,
  }),
);
