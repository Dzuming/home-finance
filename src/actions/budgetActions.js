import * as types from './actionTypes';
import { createAction } from 'redux-actions';
import env from '../../environments/config';

export const fetchBudget = () => (dispatch, getState) => {
  const { id } = getState().user;
  return dispatch({
    type: types.API_REQUEST_GET,
    payload: {
      url: `${env.api_url}/api/budget/${id}`,
      success: setBudget,
    },
  });
};

export const setBudget = createAction(types.BUDGET_CRUD.READ);
