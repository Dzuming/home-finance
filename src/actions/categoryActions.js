import * as types from './actionTypes';
import env from '../../environments/config';
import { createAction } from 'redux-actions';

export const fetchCategories = createAction(types.API_REQUEST_GET, () =>  ({
  url: `${env.api_url}/api/categories`,
  success: setCategories,
}));

export const setCategories = createAction(types.SET_CATEGORIES);

export const selectCategoryType = createAction(
  types.CATEGORIES_DRAG.SELECT,
  type => ({
    type,
  }),
);

export const resetDraggedCategories = createAction(
  types.CATEGORIES_DRAG.RESET,
  categories => ({
    categories,
  }),
);
