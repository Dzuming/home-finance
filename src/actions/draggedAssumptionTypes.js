import {createAction} from "redux-actions";
import * as types from "./actionTypes";

export const reduceAssumptionTypes = createAction(
  types.ASSUMPTION_TYPES_DRAG.REDUCE,
  type =>  ({
    type,
  }),
);

export const resetDraggedAssumptionTypes = createAction(
  types.ASSUMPTION_TYPES_DRAG.RESET,
);
