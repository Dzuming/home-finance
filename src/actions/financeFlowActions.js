import * as types from './actionTypes';
import env from '../../environments/config';

function successFinanceFlow(data) {
  return {type: types.SUCCESS_FINANCE_FLOW, data};
}

function fetchFinanceFlow(userId, selectedDate) {
  const url = `${env.api_url}/Spending/${userId}/${selectedDate}`;
  return dispatch => {
    return fetch(url, {method: 'GET'}).then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    }).then((data) => {
      dispatch(successFinanceFlow(data));
    }).catch((error => error));
  };
}
export function getFinanceFlow() {
  // Note that the function also receives getState() which lets you choose what to
  // dispatch next. This is useful for avoiding a network request if a cached
  // value is already available.
  return (dispatch, getState) => {

    const userId = getState().user.id;
    const selectedDate = getState().financeFlow.selectedDate;
    // Dispatch a thunk from thunk!
    return dispatch(fetchFinanceFlow(userId, selectedDate));
  };
}
export function setDate(date) {
  return {type: types.SET_DATE, date};
}
