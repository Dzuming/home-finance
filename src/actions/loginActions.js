import * as types from './actionTypes';
import { getUser } from './userActions';
import env from '../../environments/config';

function requestLogin () {
  return {type: types.REQUEST_LOGIN};
}

function successLogin (accessToken) {
  setStorageToken(accessToken);
  return {type: types.SUCCESS_LOGIN};
}

function loginError (message) {
  return {type: types.LOGIN_FAILURE, message};
}

function authUser () {
  return {type: types.AUTH_USER};
}

function authenticateUser (credentials) {
  const user = credentials;
  let formData = new FormData();
  formData.append('grant_type', 'password');
  formData.append('client_id', '2');
  formData.append('client_secret', 'tbvtGeM45Jnk8IFRTVXFiBJrySgtU7pgDehuZoKG');
  formData.append('username', user.email);
  formData.append('password', user.password);
  formData.append('scope', '*');
  return dispatch => {
    dispatch(requestLogin());
    return fetch(`${env.api_url}/oauth/token`, {
      method: 'POST',
      body: formData
    }).then(response => {
      if (!response.ok) {
        throw Error(response.statusText);
      }
      return response.json();
    }).then((data) => {
      const accessToken = data.access_token;
      dispatch(successLogin(accessToken));
      dispatch(authUser());
      dispatch(getUser(user.email, accessToken));
    }).catch((error => dispatch(loginError(error))));
  };
}

function setStorageToken (token) {
  localStorage.setItem('token', token);
}

export function login (data) {
  // Note that the function also receives getState() which lets you choose what to
  // dispatch next. This is useful for avoiding a network request if a cached
  // value is already available.
  return (dispatch) => {
    // Dispatch a thunk from thunk!
    return dispatch(authenticateUser(data));
  };
}
