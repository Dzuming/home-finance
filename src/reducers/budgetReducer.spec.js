import deepFreeze from 'deep-freeze';
import reducer from './budgetReducer';
import { setBudget } from '../actions/budgetActions';

const initialState = deepFreeze(reducer({}, { type: 'INIT' }));
const budget = {
  budget: 50000,
};

describe('budget reducer', () => {
  it('should handle unknown actions', () => {
    expect(reducer(initialState, { type: 'FAKE' })).toBe(initialState);
  });

  describe('get action', () => {
    it('should get budget', () => {
      expect(reducer(initialState, setBudget(budget))).toMatchSnapshot();
    });
  });
});
