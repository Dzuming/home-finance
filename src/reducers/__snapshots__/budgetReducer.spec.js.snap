// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`budget reducer get action should get budget 1`] = `50000`;
